### ISSUE TITLE
DeepPartial doesn't work with Element

### ISSUE THREAD

**Issue by TazmanianD** [krzkaczor/ts-essentials#95]
For some reason, DeepPartial doesn't seem to like the Element interface (or any of its subclasses). We get an error assigning values that should actually be type equivalent. This might be a new issue caused by a newer TypeScript version.

I'll note that the simpler definition of DeepPartial that comes with redux does not have this issue.

Repro case (error occurs on "const partialObject =..."):
```typescript
import { DeepPartial } from 'ts-essentials';

interface SomeInterface {
  element: Element;
}

const element: Element = null;
const object = { element };
const partialObject: DeepPartial<SomeInterface> = object;
```

Spits out a massive (and incomplete) error:
```
Error:(24, 7) TS2322: Type '{ element: Element; }' is not assignable to type '{ element?: { readonly assignedSlot?: { name?: string; assignedElements?: (options?: AssignedNodesOptions) => Element[]; assignedNodes?: (options?: AssignedNodesOptions) => Node[]; addEventListener?: { ...; }; ... 240 more ...; readonly style?: { ...; }; }; ... 125 more ...; getAnimations?: () => Animation[]; }; }'.
  Types of property 'element' are incompatible.
    Type 'Element' is not assignable to type '{ readonly assignedSlot?: { name?: string; assignedElements?: (options?: AssignedNodesOptions) => Element[]; assignedNodes?: (options?: AssignedNodesOptions) => Node[]; addEventListener?: { <K extends "waiting" | ... 91 more ... | "paste">(type: K, listener: (this: HTMLSlotElement, ev: HTMLElementEventMap[K]) => any...'.
      Types of property 'assignedSlot' are incompatible.
        Type 'HTMLSlotElement' is not assignable to type '{ name?: string; assignedElements?: (options?: AssignedNodesOptions) => Element[]; assignedNodes?: (options?: AssignedNodesOptions) => Node[]; addEventListener?: { <K extends "waiting" | "error" | "abort" | ... 89 more ... | "paste">(type: K, listener: (this: HTMLSlotElement, ev: HTMLElementEventMap[K]) => any, opti...'.
          Types of property 'attributes' are incompatible.
            Type 'NamedNodeMap' is not assignable to type '{ [x: number]: { readonly localName?: string; readonly name?: string; readonly namespaceURI?: string; readonly ownerElement?: { readonly assignedSlot?: { name?: string; assignedElements?: (options?: AssignedNodesOptions) => Element[]; ... 242 more ...; readonly style?: { ...; }; }; ... 125 more ...; getAnimations?: ...'.
              Index signatures are incompatible.
                Type 'Attr' is not assignable to type '{ readonly localName?: string; readonly name?: string; readonly namespaceURI?: string; readonly ownerElement?: { readonly assignedSlot?: { name?: string; assignedElements?: (options?: AssignedNodesOptions) => Element[]; ... 242 more ...; readonly style?: { ...; }; }; ... 125 more ...; getAnimations?: () => Animation...'.
                  Types of property 'childNodes' are incompatible.
                    Type 'NodeListOf<ChildNode>' is not assignable to type '{ [x: number]: { after?: (...nodes: (string | Node)[]) => void; before?: (...nodes: (string | Node)[]) => void; remove?: () => void; replaceWith?: (...nodes: (string | Node)[]) => void; readonly baseURI?: string; readonly childNodes?: ...; ... 48 more ...; removeEventListener?: (type: string, callback: EventListener...'.
                      Index signatures are incompatible.
                        Type 'ChildNode' is not assignable to type '{ after?: (...nodes: (string | Node)[]) => void; before?: (...nodes: (string | Node)[]) => void; remove?: () => void; replaceWith?: (...nodes: (string | Node)[]) => void; readonly baseURI?: string; readonly childNodes?: { ...; }; ... 48 more ...; removeEventListener?: (type: string, callback: EventListenerOrEventLis...'.
                          Types of property 'ownerDocument' are incompatible.
                            Type 'Document' is not assignable to type '{ readonly URL?: string; readonly activeElement?: { readonly assignedSlot?: { name?: string; assignedElements?: (options?: AssignedNodesOptions) => Element[]; assignedNodes?: (options?: AssignedNodesOptions) => Node[]; ... 241 more ...; readonly style?: { ...; }; }; ... 125 more ...; getAnimations?: () => Animation[...'.
                              Types of property 'currentScript' are incompatible.
                                Type 'HTMLOrSVGScriptElement' is not assignable to type '{ async?: boolean; charset?: string; crossOrigin?: string; defer?: boolean; event?: string; htmlFor?: string; integrity?: string; noModule?: boolean; referrerPolicy?: string; src?: string; text?: string; type?: string; ... 241 more ...; readonly style?: { ...; }; } | { ...; }'.
                                  Type 'HTMLScriptElement' is not assignable to type '{ async?: boolean; charset?: string; crossOrigin?: string; defer?: boolean; event?: string; htmlFor?: string; integrity?: string; noModule?: boolean; referrerPolicy?: string; src?: string; text?: string; type?: string; ... 241 more ...; readonly style?: { ...; }; } | { ...; }'.
```

Using:
ts-essentials@3.0.2
typescript@3.6.4

**Comment by quezak** [comment]
What's your definition of `Element`? When I just paste your snipped into my vscode, it points to React's global typings, where all the `Element` interfaces are empty -- which is weird and maybe it can confuse the typing system.

This snippet is typed properly for me under TS 3.6.2 -- can you check on that TS version to see if it's dependent on TS version?
(note: I had to put something for the `element` value, since passing a `null` to a non-nullable variable is invalid)
```typescript
interface SomeInterface {
  element: Element;
}

const element: Element = { a: 1 };
const object = { element };
const partialObject: DeepPartial<SomeInterface> = object;
```

**Comment by TazmanianD** [comment]
IntelliJ seems to show two possible resolutions for `Element`. One is the empty interface in react which I don't think is the one TS is using. The other is actually part of TS. I'm seeing `Element` defined in `node_modules/typescript/lib/lib.dom.d.ts`

I did try TS 3.6.2 and I do still get the error. However I tried 3.5.3 and I do *not* get the error so it seems like a change in TS 3.6 may have caused this. I'm not sure why you don't see it on 3.6.2 like I do.

And as for the null, yeah, we've got the strict null check setting off which is what allowed my original code, but your's is fine too. Just needed some value there.

**Comment by TazmanianD** [comment]
So here's another data point. I copied the entire `lib.dom.d.ts` file into the file I was using as the repro case above. That way I could start editing the definitions to try to narrow down the culprit. What I discovered is that if I remove all properties with a type composed of `typeof globalThis` (there's only 4 references to it) that this error goes away. I wonder if your definitions here need to do something special with this special global variable.

**Comment by krzkaczor** [comment]
@TazmanianD It would be great if you could provide a simple repo with a reproduction case. With lock files committed we should be able to recreate this problem.

**Comment by TazmanianD** [comment]
I tried to repro this on a new repro and was unable to do so until I pulled in the `package.json` and `package-lock.json` file from our existing project. At that point I knew it was some dependency of ours causing it and I tried to isolate which one which didn't go so well. I eventually realized that starting with the same `package.json` but deleting `package-lock.json` and creating a whole new one from scratch, that this problem went away.

So it seems like there was some library buried as some dependency in our tree that needed to get updated that was responsible for this. So I'm going to just close this.

Coincidentally I think, we just tried upgrading to TypeScript 3.7 and the `DeepPartial` type that's defined in `redux` suddenly started reporting a similar error also on a `Element` property on of our our classes (which we're breaking the rules by having one of those in our store somewhere so we just need to stop doing that).
